The DNS search path on Windows is now restored when Telepresence quits

thallgren opened the pull request:
## Description

The DNS search path that Telepresence uses to simulate the DNS lookup functionality in the connected cluster namespace was not removed by a `telepresence quit`, resulting in connectivity problems from the workstation. Telepresence will now remove the entries that it has added to the search list when it quits.

## Checklist

 - [x] I made sure to update `./CHANGELOG.yml`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [x] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

Commit:
The DNS search path on Windows is now restored when Telepresence quits

The DNS search path that Telepresence uses to simulate the DNS lookup
functionality in the connected cluster namespace was not removed by a
`telepresence quit`, resulting in connectivity problems from the
workstation. Telepresence will now remove the entries that it has added
to the search list when it quits.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix concurrent map write fatal error.

Occurred on macOS due to use of a static map that got shared among all
reporters, but revealed a badly designed constructor for the reporter
which is refactored in this commit.

The fatal error that is fixed:

    fatal error: concurrent map writes

    goroutine 42 [running]:
    github.com/telepresenceio/telepresence/v2/pkg/client/scout.getInstallIDFromFilesystem({0x1061f0b68, 0xc0005edbc0}, 0xc0001f13b0, {0x1056c9d01, 0x3})
    	github.com/telepresenceio/telepresence/v2/pkg/client/scout/reporter.go:149 +0x7e5
    github.com/telepresenceio/telepresence/v2/pkg/client/scout.NewReporterForInstallType.func1(0xc0001f13b0)
    	github.com/telepresenceio/telepresence/v2/pkg/client/scout/reporter.go:186 +0x32

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LanceEa commented on pkg/vif/device_windows.go:
Do we have any test coverage for this behavior? I'm guessing since it is windows, we would need to have a windows CI environment setup which I'm not sure if we do.

thallgren replied:
Good point. We don't have any tests to verify that the search path is restored, and we do have a CI-environment. I'll add a test.

Commit:
Add test to verify that DNS search path is restored on quit.

This test is windows specific because:
- On linux, the DNS search path is attached to the VIF, so when the VIF
  is gone, our DNS settings are gone with it.
- On macOS, there's no DNS search path. Instead, we add domains that
  correspond to the namespaces.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LanceEa reviewed (approved):
LGTM
